File: app/src/main/java/com/twismart/wallpapershd/ui/wallpaper/WallpaperFragment.java
package com.twismart.wallpapershd.ui.wallpaper;

import android.content.Context;
import android.net.Uri;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.squareup.picasso.Picasso;
import com.twismart.wallpapershd.R;
import com.twismart.wallpapershd.data.model.Wallpaper;

public class WallpaperFragment extends Fragment {

    private final String TAG = getClass().getSimpleName();

    private static final String ARG_WALLPAPER = "wallpaper";

    private Wallpaper wallpaper = null;

    private OnFragmentInteractionListener mListener;

    private ImageView imageWallpaper;

    // Dimensions used for the wallpaper
    private float widthEnd, heightEnd;


    public WallpaperFragment() {
        // Required empty public constructor
    }

    public static WallpaperFragment newInstance(Wallpaper wallpaper) {
        WallpaperFragment fragment = new WallpaperFragment();
        Bundle args = new Bundle();
        args.putParcelable(ARG_WALLPAPER, wallpaper);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            wallpaper = getArguments().getParcelable(ARG_WALLPAPER);
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, final Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_wallpaper, container, false);

        imageWallpaper = (ImageView) v.findViewById(R.id.image_wallpaper);
        // It's necessary use post method because of other way this would cause error due to imageWallpaper.getWidth() or getHeight() == 0
        imageWallpaper.post(new Runnable() {
            @Override
            public void run() {
                try{
                    // Calculate the width and height in order to fit the image correctly
                    float widthWallpaper = Float.parseFloat(wallpaper.getWidth());
                    float heightWallpaper = Float.parseFloat(wallpaper.getHeight());

                    if(widthWallpaper > heightWallpaper){
                        widthEnd = imageWallpaper.getWidth();
                        heightEnd = (heightWallpaper / widthWallpaper) * widthEnd;
                    }
                    else {
                        heightEnd = imageWallpaper.getHeight();
                        widthEnd = (widthWallpaper / heightWallpaper) * heightEnd;
                    }
                    // Pick up whether the image comes from a url or a path
                    Log.d(TAG, "onCreate: widthEnd " + widthEnd + " heightEnd " + heightEnd);
                    String sourceImage = wallpaper.getUrlImage().isEmpty() ? wallpaper.getPathImage(): wallpaper.getUrlImage();
                    Glide.with(getContext())
                            .load(sourceImage)
                            .asBitmap()
                            .override((int)widthEnd, (int)heightEnd)
                            .diskCacheStrategy(DiskCacheStrategy.RESULT)
                            .into(imageWallpaper);
                }
                catch (Exception e){
                    e.printStackTrace();
                }
            }
        });

        return v;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnFragmentInteractionListener) {
            mListener = (OnFragmentInteractionListener) context;
        } else {
            //throw new RuntimeException(context.toString() + " must implement OnFragmentInteractionListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    public interface OnFragmentInteractionListener {
        void onFragmentInteraction(Uri uri);
    }
}

File: app/src/main/kotlin/com/twismart/wallpapershd/ui/wallpaper/WallpaperFragment.kt
package com.twismart.wallpapershd.ui.wallpaper

import android.content.Context
import android.net.Uri
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.twismart.wallpapershd.R
import com.twismart.wallpapershd.data.model.Wallpaper
import com.twismart.wallpapershd.utils.debug
import com.twismart.wallpapershd.utils.inflate
import kotlinx.android.synthetic.main.fragment_wallpaper.*


class WallpaperFragment : Fragment() {

    companion object {
        private val ARG_WALLPAPER = "wallpaper"
        fun newInstance(wallpaper: Wallpaper): WallpaperFragment {
            val fragment = WallpaperFragment()
            val args = Bundle()
            args.putParcelable(ARG_WALLPAPER, wallpaper)
            fragment.arguments = args
            return fragment
        }
    }

    lateinit private var wallpaper: Wallpaper

    private var mListener: OnFragmentInteractionListener? = null

    // Dimensions used for the wallpaper
    private var widthEnd = 0f
    private var heightEnd = 0f

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        wallpaper = arguments.getParcelable(ARG_WALLPAPER)
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return context?.inflate(R.layout.fragment_wallpaper, container)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        // It's necessary use post method because of other way this would cause error due to imageWallpaper.getWidth() or getHeight() == 0
        imageWallpaper.post {
            try {
                // Calculate the width and height in order to fit the image correctly
                val widthWallpaper = wallpaper.width.toFloat()
                val heightWallpaper = wallpaper.height.toFloat()

                if (widthWallpaper > heightWallpaper) {
                    widthEnd = imageWallpaper.width.toFloat()
                    heightEnd = heightWallpaper / widthWallpaper * widthEnd
                } else {
                    heightEnd = imageWallpaper.height.toFloat()
                    widthEnd = widthWallpaper / heightWallpaper * heightEnd
                }
                // Pick up whether the image comes from a url or a path
                debug("onCreate: widthEnd $widthEnd heightEnd $heightEnd")
                val sourceImage = if (wallpaper.urlImage.isEmpty()) wallpaper.pathImage else wallpaper.urlImage
                Glide.with(context)
                        .load(sourceImage)
                        .asBitmap()
                        .override(widthEnd.toInt(), heightEnd.toInt())
                        .diskCacheStrategy(DiskCacheStrategy.RESULT)
                        .into(imageWallpaper!!)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        if (context is OnFragmentInteractionListener) {
            mListener = context as OnFragmentInteractionListener?
        } else {
            //throw new RuntimeException(context.toString() + " must implement OnFragmentInteractionListener");
        }
    }

    override fun onDetach() {
        super.onDetach()
        mListener = null
    }

    interface OnFragmentInteractionListener {
        fun onFragmentInteraction(uri: Uri)
    }
}